### PR TITLE
use a cached version of site locale, to translate shutdown messages

### LINKED ISSUE #12986: Exception, when stopping instance (^C)
**Describe the bug**
When closing an instance, it's giving an exception since a little while before 0.36.0, but I thought it was my build environment.

**To Reproduce**
1. Start with H2 backend `java -jar metabase.jar`
2. Don't use instance for a few minutes (perhaps 10min)
3. Stop process with `^C` (Ctrl+C)
4. Shows exception:
```
07-23 15:34:23 DEBUG middleware.log :: GET /api/dashboard/6 200 4.5 ms (4 DB calls) App DB connections: 0/10 Jetty threads: 4/50 (5 idle, 0 queued) (104 total active threads) Queries in flight: 0 (0 queued)
^CException in thread "Thread-10" org.h2.jdbc.JdbcSQLException: Database is already closed (to disable automatic closing at VM shutdown, add ";DB_CLOSE_ON_EXIT=FALSE" to the db URL) [90121-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.message.DbException.get(DbException.java:144)
	at org.h2.jdbc.JdbcConnection.checkClosed(JdbcConnection.java:1526)
	at org.h2.jdbc.JdbcConnection.checkClosed(JdbcConnection.java:1502)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:302)
	at com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:387)
	at clojure.java.jdbc$prepare_statement.invokeStatic(jdbc.clj:679)
	at clojure.java.jdbc$prepare_statement.invoke(jdbc.clj:626)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1112)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at toucan.db$query.invokeStatic(db.clj:288)
	at toucan.db$query.doInvoke(db.clj:284)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at toucan.db$simple_select.invokeStatic(db.clj:394)
	at toucan.db$simple_select.invoke(db.clj:383)
	at toucan.db$simple_select_one.invokeStatic(db.clj:420)
	at toucan.db$simple_select_one.invoke(db.clj:409)
	at toucan.db$select_one.invokeStatic(db.clj:627)
	at toucan.db$select_one.doInvoke(db.clj:620)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at toucan.db$select_one_field.invokeStatic(db.clj:636)
	at toucan.db$select_one_field.doInvoke(db.clj:629)
	at clojure.lang.RestFn.invoke(RestFn.java:442)
	at metabase.models.setting.cache$cache_out_of_date_QMARK_.invokeStatic(cache.clj:98)
	at metabase.models.setting.cache$cache_out_of_date_QMARK_.invoke(cache.clj:79)
	at metabase.models.setting.cache$restore_cache_if_needed_BANG_.invokeStatic(cache.clj:151)
	at metabase.models.setting.cache$restore_cache_if_needed_BANG_.invoke(cache.clj:128)
	at metabase.models.setting$db_or_cache_value.invokeStatic(setting.clj:175)
	at metabase.models.setting$db_or_cache_value.invoke(setting.clj:169)
	at metabase.models.setting$get_string.invokeStatic(setting.clj:188)
	at metabase.models.setting$get_string.invoke(setting.clj:178)
	at clojure.lang.Var.invoke(Var.java:384)
	at metabase.util.i18n.impl$fn__7770$f__7771.invoke(impl.clj:156)
	at metabase.util.i18n.impl$site_locale_from_setting.invokeStatic(impl.clj:163)
	at metabase.util.i18n.impl$site_locale_from_setting.invoke(impl.clj:160)
	at metabase.util.i18n$site_locale.invokeStatic(i18n.clj:31)
	at metabase.util.i18n$site_locale.invoke(i18n.clj:28)
	at metabase.util.i18n$translate_site_locale.invokeStatic(i18n.clj:52)
	at metabase.util.i18n$translate_site_locale.doInvoke(i18n.clj:49)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.util.i18n.SiteLocalizedString.toString(i18n.clj:75)
	at clojure.core$str.invokeStatic(core.clj:553)
	at clojure.core$str.invoke(core.clj:544)
	at metabase.core$destroy_BANG_.invokeStatic(core.clj:47)
	at metabase.core$destroy_BANG_.invoke(core.clj:44)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
5. Starting an instance, but without inactivity, and then shutting down just shows the regular messages:
```
07-23 15:53:07 DEBUG middleware.log :: GET /api/automagic-dashboards/database/4/candidates 200 671.4 ms (6 DB calls) App DB connections: 0/4 Jetty threads: 4/50 (3 idle, 0 queued) (47 total active threads) Queries in flight: 0 (0 queued)
^C07-23 15:53:11 INFO metabase.core :: Metabase Shutting Down ...
07-23 15:53:11 INFO metabase.server :: Shutting Down Embedded Jetty Webserver
07-23 15:53:11 INFO metabase.core :: Metabase Shutdown COMPLETE
```

**Information about your Metabase Installation:**
```
07-23 15:50:42 INFO metabase.core :: Starting Metabase version v0.36.0 (b31049f release-0.36.x) ...
07-23 15:50:42 INFO metabase.core :: System info:
 {"file.encoding" "UTF-8",
 "java.runtime.name" "OpenJDK Runtime Environment",
 "java.runtime.version" "11.0.8+10",
 "java.vendor" "AdoptOpenJDK",
 "java.vendor.url" "https://adoptopenjdk.net/",
 "java.version" "11.0.8",
 "java.vm.name" "OpenJDK 64-Bit Server VM",
 "java.vm.version" "11.0.8+10",
 "os.name" "Linux",
 "os.version" "5.3.0-62-generic",
 "user.language" "en",
 "user.timezone" "Europe/Copenhagen"}
```

**Severity**
Not sure if this might cause some problem in the future, but I haven't noticed anything, so just giving P3.

### PR DESCRIPTION
**Pull request by escherize**
fixes #12986

Pressing Ctrl-C to shutdown the jar can throw a stacktrance

- This happened because the db is closed before the call to `destroy!`

- it occurs when translating the two shutdown messages. So caching the
  site locale and using it to translate those messages makes it work.

- testing this requires starting and `sigint`-ing it, so to test, build a
  jar with this, let it's initialization complete, and hit Ctrl-C. You
  should see: two messages about shutdown (translated into whatever
  locale you have set) and NO stacktrace.

```shell
clj -T:build uberjar
cp target/uberjar/metabase.jar <your-temp-location>
```

AFTER:
```shell
MB_SITE_LOCALE=fr MB_JETTY_PORT=3005 java -jar metabase.jar
...snip ...
^C2022-04-20 16:12:45,481 INFO metabase.core :: Metabase en cours darrêt...
2022-04-20 16:12:45,483 INFO metabase.server :: Arrêt du serveur web Jetty intégré
2022-04-20 16:12:45,512 INFO metabase.core :: Arrêt de Metabase EFFECTIF
```

BEFORE:
```shell
... snip ...
07-23 15:34:23 DEBUG middleware.log :: GET /api/dashboard/6 200 4.5 ms (4 DB calls) App DB connections: 0/10 Jetty threads: 4/50 (5 idle, 0 queued) (104 total active threads) Queries in flight: 0 (0 queued)
^CException in thread "Thread-10" org.h2.jdbc.JdbcSQLException: Database is already closed (to disable automatic closing at VM shutdown, add ";DB_CLOSE_ON_EXIT=FALSE" to the db URL) [90121-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.message.DbException.get(DbException.java:144)
	at org.h2.jdbc.JdbcConnection.checkClosed(JdbcConnection.java:1526)
	at org.h2.jdbc.JdbcConnection.checkClosed(JdbcConnection.java:1502)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:302)
	at com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:387)
	at clojure.java.jdbc$prepare_statement.invokeStatic(jdbc.clj:679)
	at clojure.java.jdbc$prepare_statement.invoke(jdbc.clj:626)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1112)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at toucan.db$query.invokeStatic(db.clj:288)
	at toucan.db$query.doInvoke(db.clj:284)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at toucan.db$simple_select.invokeStatic(db.clj:394)
	at toucan.db$simple_select.invoke(db.clj:383)
	at toucan.db$simple_select_one.invokeStatic(db.clj:420)
	at toucan.db$simple_select_one.invoke(db.clj:409)
	at toucan.db$select_one.invokeStatic(db.clj:627)
	at toucan.db$select_one.doInvoke(db.clj:620)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at toucan.db$select_one_field.invokeStatic(db.clj:636)
	at toucan.db$select_one_field.doInvoke(db.clj:629)
	at clojure.lang.RestFn.invoke(RestFn.java:442)
	at metabase.models.setting.cache$cache_out_of_date_QMARK_.invokeStatic(cache.clj:98)
	at metabase.models.setting.cache$cache_out_of_date_QMARK_.invoke(cache.clj:79)
	at metabase.models.setting.cache$restore_cache_if_needed_BANG_.invokeStatic(cache.clj:151)
	at metabase.models.setting.cache$restore_cache_if_needed_BANG_.invoke(cache.clj:128)
	at metabase.models.setting$db_or_cache_value.invokeStatic(setting.clj:175)
	at metabase.models.setting$db_or_cache_value.invoke(setting.clj:169)
	at metabase.models.setting$get_string.invokeStatic(setting.clj:188)
	at metabase.models.setting$get_string.invoke(setting.clj:178)
	at clojure.lang.Var.invoke(Var.java:384)
	at metabase.util.i18n.impl$fn__7770$f__7771.invoke(impl.clj:156)
	at metabase.util.i18n.impl$site_locale_from_setting.invokeStatic(impl.clj:163)
	at metabase.util.i18n.impl$site_locale_from_setting.invoke(impl.clj:160)
	at metabase.util.i18n$site_locale.invokeStatic(i18n.clj:31)
	at metabase.util.i18n$site_locale.invoke(i18n.clj:28)
	at metabase.util.i18n$translate_site_locale.invokeStatic(i18n.clj:52)
	at metabase.util.i18n$translate_site_locale.doInvoke(i18n.clj:49)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.util.i18n.SiteLocalizedString.toString(i18n.clj:75)
	at clojure.core$str.invokeStatic(core.clj:553)
	at clojure.core$str.invoke(core.clj:544)
	at metabase.core$destroy_BANG_.invokeStatic(core.clj:47)
	at metabase.core$destroy_BANG_.invoke(core.clj:44)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### CODE REVIEW

**Comment by ariya** [comment]
@escherize Looks like this breaks launching the Uberjar at initialization.

To reproduce: `./bin/build` and then run it `java -jar target/uberjar/metabase.jar`:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:416)
        at clojure.lang.RT.classForName(RT.java:2212)
        at clojure.lang.RT.classForName(RT.java:2221)
        at clojure.lang.RT.loadClassForName(RT.java:2240)
        at clojure.lang.RT.load(RT.java:449)
        at clojure.lang.RT.load(RT.java:424)
        at clojure.core$load$fn__6856.invoke(core.clj:6115)
        at clojure.core$load.invokeStatic(core.clj:6114)
        at clojure.core$load.doInvoke(core.clj:6098)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.lang.Var.invoke(Var.java:384)
        at clojure.lang.Util.loadWithClass(Util.java:251)
        at metabase.core.<clinit>(Unknown Source)
Caused by: java.lang.IllegalStateException: Attempting to call unbound fn: #'metabase.util.i18n.impl/locale
        at clojure.lang.Var$Unbound.throwArity(Var.java:45)
        at clojure.lang.AFn.invoke(AFn.java:32)
        at clojure.lang.Var.invoke(Var.java:384)
        at clojure.lang.LispReader$CtorReader.readTagged(LispReader.java:1448)
        at clojure.lang.LispReader$CtorReader.invoke(LispReader.java:1427)
        at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:846)
        at clojure.lang.LispReader.read(LispReader.java:285)
        at clojure.lang.LispReader.read(LispReader.java:216)
        at clojure.lang.LispReader.read(LispReader.java:205)
        at clojure.lang.RT.readString(RT.java:1879)
        at clojure.lang.RT.readString(RT.java:1874)
        at metabase.core__init.__init0(Unknown Source)
        at metabase.core__init.<clinit>(Unknown Source)```

**Comment by ariya** [comment]
@escherize can you rebase against the latest master? That likely gets rid of the flaky viz tests.

**Comment by dpsutton** [comment]
@ariya I cannot reproduce your experience. `bin/build` and then `MB_JETTY_PORT=3006 java -jar target/uberjar/metabase.jar` gives me a pleasant shutdown and no errors at startup:

```
2022-04-25 09:44:49,104 INFO metabase.core :: Metabase Initialization COMPLETE
^C2022-04-25 09:45:26,333 INFO metabase.core :: Metabase Shutting Down ...
2022-04-25 09:45:26,344 INFO metabase.server :: Shutting Down Embedded Jetty Webserver
2022-04-25 09:45:26,426 INFO metabase.core :: Metabase Shutdown COMPLETE
```
(note the `^C`).

and then `MB_SITE_LOCALE=fr MB_JETTY_PORT=3006 java -jar target/uberjar/metabase.jar`

```
2022-04-25 09:50:25,201 INFO metabase.core :: Utilisez sil vous plaît lURL suivante pour configurer votre installation Metabase :

http://localhost:3006/setup/


2022-04-25 09:50:25,214 INFO metabase.core :: Initialisation de Metabase EFFECTIVE
^C2022-04-25 09:50:28,305 INFO metabase.core :: Metabase en cours darrêt...
2022-04-25 09:50:28,319 INFO metabase.server :: Arrêt du serveur web Jetty intégré
2022-04-25 09:50:28,370 INFO metabase.core :: Arrêt de Metabase EFFECTIF
```

**Comment by ariya** [comment]
> @ariya I cannot reproduce your experience.

I think after commit `413d8677`, the startup crash was fixed.

**Comment by escherize** [comment]
I wanted to put in a tagged reader `#locale "en"` however it seems like the reader hadn't loaded `metabase.util.i18n.impl/locale` yet. I considered requiring it in the core namespace but it felt more confusing than just using the (perfectly valid) string form "en"

File: src/metabase/core.clj [hunk 1]
```diff
@@ -94,6 +102,7 @@
   ;; startup database.  validates connection & runs any necessary migrations
   (log/info (trs "Setting up and migrating Metabase DB. Please sit tight, this may take a minute..."))
   (mdb/setup-db!)
+  (reset! cached-shutdown-locale (i18n/site-locale))
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/core.clj:105]: That's possible. One thing we noticed was that `destroy!` doesn't tear down the db connections. So we're not sure why they are gone already in `destroy!`. But by the time the `destroy!` function runs the database is gone so it's not clear who would pass that to the function. But as to your point in general, it's two logged messages saying "Metabase Shutting Down ..." and then "Metabase Shutdown COMPLETE". I'd rather be very safe than flirt with a race during shutdown.

[the remark this answers is not in the document]
Comment (escherize) [reply src/metabase/core.clj:105]: Wondering if the increased confusion vs. payoff (that the ^C message is in your newly updated language) is worth it

**Comment by escherize** [comment]
I found a better place to keep an up-to-date copy of the site-locale in the event of a db disconnection. Will have an update out shortly

**Comment by escherize** [comment]
Here's the output from hitting `^C`:

```

... snip ...
/50 (2 idle, 0 queued) (78 total active threads) Queries in flight: 0 (0 queued); h2 DB 1 connections: 0/1 (0 threads blocked)
2022-04-26 19:54:23,190 INFO metabase.core :: Metabase Shutting Down ...
2022-04-26 19:54:23,205 INFO metabase.server :: Shutting Down Embedded Jetty Webserver
2022-04-26 19:54:23,271 INFO metabase.core :: Metabase Shutdown COMPLETE
```

**Review by dpsutton — APPROVED** [review]

**Pull request merged**